docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+
+This document collects proposals made by various contributors (see [references](#references)) and presents a simplified serial class hierarchy. The aim is to offer two public classes, one for unbuffered I/O and the other for buffered I/O, with a clear definition of which one should be used when. All other serial classes are deprecated and should be removed at a later stage. In addition, the serial classes dependencies will be optimised to only include components that provide valuable functionality.
+
+#### Serial classes in Mbed OS 5.14
+
+**Serial**
+
+`Serial` provides unbuffered I/O and is using blocking HAL calls. It only works for simple printing. 
+
+Its main limitations are:
+- It uses a mutex lock so it cannot be used from interrupts. 
+- It wastes a lot of CPU time spinning waiting for serial output and it cannot be used reliably for input because it needs interrupt-speed response on read calls but its high-level API cannot be used from interrupts.
+- It lacks buffering so it cannot be used reliably for input and output without flow control.
+- It pulls in the C library stdio system because it uses `Stream`.
+
+**UARTSerial**
+
+`UARTSerial` provides buffered I/O. It can be used reliably for input from non-interrupt context and to avoid excess spinning waiting for transmission buffer space. It also does better blocking and correct POSIX-like `FileHandle` semantics. 
+
+It does not use `Stream` so it has the advantage of not pulling in the C stdio library. The drawback is that it loses the ability to call printf directly.
+
+**RawSerial**
+
+`RawSerial` provides unbuffered I/O. It has no locks so it is safe to use in interrupts provided that it is the only instance that is using that serial port.
+
+### Requirements and assumptions
+
+None
+
+# System architecture and high-level design
+
+Applications will have a choice between two public serial classes: 
+- `BufferedSerial` for a normal application
+- `UnbufferedSerial` for an interrupt driven application or one that needs to have more control
+
+### System architecture and component interaction
+
+The below diagram shows the inheritance hierarchy for the serial classes.
+
+![Serial classes](./diagrams/serial-classes.png)
+
+`MinimalSerial` is an abstract class that extracts basics I/O functionality from `SerialBase` so it can be re-used for a minimal serial console in the retarget code.
+
+# Detailed design
+
+### Detailed design : BufferedSerial
+
+`BufferedSerial` is `UARTSerial` renamed to convey the original purpose of the class. In addition, the new class will allow configurability of the transmit and receive buffer size so that each instance of the class can configure its buffer size according to its intended usage, for example a modem may need a large input buffer and a console may not need any input buffer. 
+
+The below diagram shows the detailed inheritance hierarchy for `BufferedSerial`.
+
+![BufferedSerial class](./diagrams/buffered-serial.png)
```

Comment:
I'd suggest annotating the `BufferedSerial`->`SerialBase` inheritance somehow to note that it's private. As `SerialBase` isn't polymorphic, that forces people to via the buffer. If the `SerialBase` was public, people could bypass the buffer, which would be weird.